### ISSUE TITLE
(Re)Compression of existing repository?

### ISSUE THREAD

**Issue by oderwat** [kopia/kopia#728]
Is there a way to set compression after taking snapshots already and then recompress the repository?

And probably related: How does deduplication work when compression was enabled later? Will new files be re-uploaded compressed or matched with blocks that are already in the repository?

I find it hard to find any information about how compression works.

**Comment by julio-lopez** [comment]
Compression is policy based. It can be enabled after having taken snapshots, and can also be set at the subdirectory granularity. However, the setting is not applied retroactively. Existing snapshots remain uncompressed.

For snapshots taken after compression has been enabled, completely unmodified files (the file paths and file times match those of the last snapshot) will not be re-uploaded. New files that end up being compressed are uploaded, even if their uncompressed contents match those of existing files.

I hope this helps.

**Comment by oderwat** [comment]
I still don't understand the process, really. From what you write, it sounds like the compression is basically done per file and before deduplication.

Wouldn't this harm the possibilities of deduplication? I usually use `gzip --rsyncable` with other deduplicating storage tools. Which works very well with tar gzip piped data.

You also write they are re-uploaded even if their content matches. Let's take a big file (for example a MySQL database binary data or index file), which is modified somewhere in between or at least growing all the time. It sounds as if enabling compression will re-upload this whole file again (and again?) because it now looks different every time (because of the compression).

Is compression basically just a "preprocessing" step as if one would compress the files beforehand?

I expected that the data gets compressed after the deduplication as an additional way to save storage and bandwidth. 
Basically, the blobs of data are compressed before transmission and inside the repository, but their hashes refer to the uncompressed data.

**Comment by jkowalski** [comment]
objects are written in the following stages:

splitter -> compressor -> hash/dedup -> encrypt -> package

Assuming compressors produce stable outputs (and they usually do), deduplication property is preserved.

We could consider compressing between hash and encrypt, although that would require minor index format change (we'd need to remember which encrypted blocks are also compressed and which ones are not, that is just one bit) but would potentially allow re-compression of repositories.

Not sure how compatible that would be with directory-level policies, but perhaps worth trying at some point.

**Comment by oderwat** [comment]
I don't think that compressors are mostly stable. I think it is the other way around and most produce very different results for minimal file changes. But I may be wrong, especially here, because I actually don't know how the splitter works and how big the blocks are. At least I god terrible results with anything but `gzip --rsyncable` for something like a MySQL dump or, in my use case, the archive dump of a fairly big MongoDB database. Later one still needs Restic because of the missing stdio functionality from Kopia.

I think the directory-level policies are not that important. At first, I thought: Oh, how cool.. I can have some directories being compressed and also chose how hard it should try. But now I realized that I would much more prefer that the blobs in the repository use as little space as possible and esp. are transmitted "cheap" while still getting whole snapshot deduplication. I like the idea of creating a fast backup of a snapshot and then have some other tool compress data in the repository when needed. This could also implement the bitrot protection others asked for.

P.S.: I am not sure if somebody mentioned it yet, but one killer feature would be a restoration that can utilize locally available data. Kinda like "check what's there; fill the gaps." I would love to restore some 25 GB directory with big files where just a few changes are in new snapshots. Kinda like a "repository based sync of local files."

**Comment by jkowalski** [comment]
Interesting perspective. Are you saying regular buzhash-based splitter is not doing a good when splitting a MySQL dump?

Can you try an experiment?

1. Take two, relatively similar but not the same snapshots of large (100MB+) MySQL DB (does not really matter whether compressed or not)
2. Find object IDs of files in question (they will start with `Ix<hash>`). You can use `kopia ls -l` for that.
3. Strip `I` prefix and do `kopia show x<hash>` on both files and compare the resulting JSON - they should be very similar to each other and share most of object IDs 

You can then individually examine objects that differ to see how Kopia has split and stored the large file and how many bytes were/were not deduplicated.

Each of constituting objects (listed as `o` in the index) is individually compressed. You should be able to compare the length of the object (`l` field) with the size of the underlying content (`o` without `Z` prefix)

```
$ kopia content ls --prefix CONTENT_ID -l                                                                                                                                          
```

The output will be:

```
CONTENT_ID DATE PACK OFFSET+STORED_SIZE
```

**Comment by oderwat** [comment]
Will try soon!

**Comment by oderwat** [comment]
`kopia ls -l` does not show any lines starting with `Ix` only `k` and `x`. 

But I read some more and ended up there: https://moinakg.wordpress.com/2013/06/22/high-performance-content-defined-chunking/ 

What was confusing me is thinking that the "deduplication" contains what the "chunking" does. I think this comes down to the difference in running compression before or after the splitter. With something like `gzip --rsyncable`, you seem to get away with compression before splitting. I guess what actually confused me was that Restic does not have built-in compression, so I needed to find a way to do it myself.

My original use case / problem is all described by this line:

```
mongodump --db=sampledb --archive -u=sampleuser -p=samplepw | gzip --rsyncable -c | restic.sh backup --stdin-filename mongodb-sampledb.archive --tag bak_sampledb --stdin.`
```

The dump greatest a single 12 GB archive file, which is piped through gzip and then into restic. Using gzip here shrinks the actual data processed by restic to about 2 GB. When using this without the rsyncable flag, the data is not well deduplicated.

So besides that, Kopia does not yet have stdio possibilities. It actually would compress after the splitter step, and as far as I understand it, the rolling hash will speed up the operation, and the compression afterward will compress the data.

I hope I didn't waste your time and understanding it better now.
